### ISSUE TITLE
Efficient Bulk Create/Insert With One Database Call

### ISSUE THREAD

**Issue by shadyanwar** [loopbackio/loopback-next#3357]
## Suggestion

The [createAll()](https://loopback.io/doc/en/lb4/apidocs.repository.defaultcrudrepository.createall.html) method provides the ability to bulk create records, however, it is translated into multiple insert statements since the [dao](https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/dao.js) layer inserts records one by one. I suggested that, in MySQL for example, LB4 uses the [multiple-row INSERT syntax](https://dev.mysql.com/doc/refman/5.5/en/optimizing-innodb-bulk-data-loading.html) which aims at reducing the communication overhead between the client and the server.

## Use Cases
Less server overhead and better performance. In case of inserting around 10k records, the performance impact due to the current implementation is huge. A query might take something like 10 seconds or more and raise the server utilization extensively. Whereas with the multiple-row INSERT syntax the same query could take something like 1 second.

## Examples
Optimally, instead of:
```
INSERT INTO `tbl`(`fld1`,`fld2`,`fld3`,`fld4`) VALUES(1,2,3,4);
INSERT INTO `tbl`(`fld1`,`fld2`,`fld3`,`fld4`) VALUES(10,20,30,40);
INSERT INTO `tbl`(`fld1`,`fld2`,`fld3`,`fld4`) VALUES(100,200,300,400);
```

We would have:
```INSERT INTO `tbl`(`fld1`,`fld2`,`fld3`,`fld4`) VALUES(1,2,3,4),(10,20,30,40),(100,200,300,400);```

## Acceptance criteria

TBD - will be filled by the team.

**Comment by pankajcheema** [comment]
I am agree with the point of @shadyanwar . I am also looking for the same .

**Comment by pankajcheema** [comment]
any update on the same ?

**Comment by shadyanwar** [comment]
@dhmlau I’m happy to help

**Comment by dishantr16** [comment]
I'm using Mongodb facing the same issue.... have any one found the solution?

**Comment by samarpanB** [comment]
@dhmlau @achrinza This is a very significant issue. We fixed it by overriding the createAll method in repository. But that's not right, because I have to do this at multiple places. I can create a mixin and do this too, but since it might be a performance bug for most of the databases, should we not have it in LB4 itself ?

Having said that, I tried to look for a solution to this in connector level. I checked [postgres connector](https://github.com/loopbackio/loopback-connector-postgresql/blob/master/lib/postgresql.js), for instance, but there was no reference of createAll here. Then I checked its base class in [loopback-connector](https://github.com/loopbackio/loopback-connector/blob/master/lib/sql.js). There also I can only see [create](https://github.com/loopbackio/loopback-connector/blob/6785802301fa39ab0d1dab93ad608d3cad413083/lib/sql.js#L620) method. I checked juggler-bridge then. There I saw the createAll method making an internal call to create method. The create method invokes modelclass.create, which eventually invokes the connector create method. See [below](https://github.com/loopbackio/loopback-next/blob/e6eacffa3b36c1769861ad5a4db71c615c17b1f3/packages/repository/src/repositories/legacy-juggler-bridge.ts#L497).

```
async createAll(entities: DataObject<T>[], options?: Options): Promise<T[]> {
    // perform persist hook
    const data = await Promise.all(
      entities.map(e => this.entityToData(e, options)),
    );
    const models = await ensurePromise(this.modelClass.create(data, options));
    return this.toEntities(models);
  }
```

I am basically trying to raise a PR to have this capability of either using create method multiple times or using insert query for multiple records. This definitely should go to connectors actually. Can you confirm, I have to modify the juggler bridge function first and then move to connectors, or am I getting it wrong ?

**Comment by achrinza** [comment]
A summary of the code flow is:

1. LB4 Repository calls Juggler Model Class `.create`
2. Juggler Model Class gets the above method (and others) from being [mixin-ed](https://github.com/loopbackio/loopback-datasource-juggler/blob/f8d7ca94740887e9beefc586442f025c12c29e7f/lib/datasource.js#L742) Juggler [DataAccessObject](https://github.com/loopbackio/loopback-next/blob/e6eacffa3b36c1769861ad5a4db71c615c17b1f3/packages/repository/src/repositories/legacy-juggler-bridge.ts#L497) by default
3. Juggler DataAccessObject `.create` calls the Juggler SQL Connectors' `.create`
4. The Juggler SQL Connectors inherit from `loopback-connector`'s [`SQLConnector`](https://github.com/loopbackio/loopback-connector/blob/6785802301fa39ab0d1dab93ad608d3cad413083/lib/sql.js#L620)

So:

1.  "crux" of the issue is that the connector needs to implement `.createAll`.
2. Since the SQL Connectors inherit from a common `SQLConnector`, updating that should be sufficient (assuming the SQL dialect isn't too different).
3. The Juggler DataAccessObject also needs to be updated to implement `.createAll`, which would be mixin-ed into the Juggler Model Class during runtime
4. Finally, LB4 DefaultCrudRepository's `.createAll` needs to be updated in a backwards-compatible way.

**Comment by samarpan-b** [comment]
Thanks @achrinza . However, I think mongodb connector also uses the same sql connector. Mongo DB has different query for insert multiple. I'll check and confirm.

**Comment by achrinza** [comment]
@samarpan-b The MongoDB connector inherits from `Connector` instead of `SQLConnector`:

https://github.com/loopbackio/loopback-connector-mongodb/blob/51d132a48c0b559978a3d379a9695ff6bc1d888a/lib/mongodb.js#L211

Though it does implement the necessary CRUD functions to be compatible with Juggler DataAccessObject.

**Comment by samarpanB** [comment]
@achrinza I checked if the multiple insert in one query dialect is available in different databases or not and also if its available, is it same or not ?
I observed that only [postgresql](https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-insert-multiple-rows/), [mysql](https://www.mysqltutorial.org/mysql-insert-multiple-rows/), [sql server](https://www.sqlservertutorial.net/sql-server-basics/sql-server-insert-multiple-rows/) have a common dialect. [Oracle](https://livesql.oracle.com/apex/livesql/file/content_BM1LJQ87M5CNIOKPOWPV6ZGR3.html) supports multiple insert in one query but dialect is different and cassandra doesn't even support it.

**Comment by samarpanB** [comment]
Point 1 to 5 are done and PRs are raised. @achrinza please do take a look.

**Comment by samarpanB** [comment]
Point 6 is also done. But its in draft. We need connector and juggler to release first. Then only I can update dependency in @loopback/repository.

**Comment by samarpanB** [comment]
This is now released for postgresql in version 6.0.0 of loopback-connector-postgresql. See here - https://github.com/loopbackio/loopback-connector-postgresql/blob/master/CHANGES.md
